src/tasks/PageIterator.ts

```diff
@@ -0,0 +1,332 @@
+      error.name = "Invalid Page Result Error";
+      throw error;
+    }
+    if (!callback) {
+      const error = new Error("Callback is undefined, Please provide a valid callback");
+      error.name = "Invalid Callback Error";
+      throw error;
+    }
+    if (!parsableFactory) {
+      const error = new Error("Parsable factory is undefined, Please provide a valid parsable factory");
+      error.name = "Invalid Parsable Factory Error";
+      throw error;
+    }
+    if (!errorMappings) {
+      const error = new Error("Error mappings is undefined, Please provide a valid error mappings");
+      error.name = "Invalid Error Mappings Error";
+      throw error;
+    }
+    this.requestAdapter = requestAdapter;
+    this.currentPage = pageResult;
+
+    this.cursor = 0;
+    this.complete = false;
+    this.errorMappings = errorMappings;
+    this.parsableFactory = parsableFactory;
+    this.callback = callback;
+
+    if (!options) {
+      options = {};
+    }
+    if (!options.headers) {
+      options.headers = new Headers();
+    }
+    if (!options.headers.has("Content-Type")) {
+      options.headers.add("Content-Type", "application/json");
+    }
+    this.options = options;
+    this.pagingState = "NotStarted";
+  }
+
+  /**
+   * @public
+   * Getter to get the deltaLink in the current response
+   * @returns A deltaLink which is being used to make delta requests in future
+   */
+  public getOdataDeltaLink(): string | undefined {
+    const deltaLink = this.currentPage?.["@odata.deltaLink"] as string | undefined;
+    return this.currentPage?.odataDeltaLink ?? deltaLink;
+  }
+
+  /**
+   * @public
+   * Getter to get the nextLink in the current response
+   * @returns A nextLink which is being used to make requests in future
+   */
+  public getOdataNextLink(): string | undefined {
+    const nextLink = this.currentPage?.["@odata.nextLink"] as string | undefined;
+    return this.currentPage?.odataNextLink ?? nextLink;
+  }
+
+  /**
+   * @public
+   * @async
+   * Iterates over the collection and kicks callback for each item on iteration. Fetches next set of data through nextLink and iterates over again
+   * This happens until the nextLink is drained out or the user responds with a red flag to continue from callback
+   */
+  public async iterate() {
+    const keepIterating = true;
+
+    while (keepIterating) {
+      const advance = this.enumerate();
+      if (!advance) {
+        return;
+      }
+
+      const nextLink = this.getOdataNextLink();
+      if (
+        (nextLink === undefined || nextLink === null || nextLink === "") &&
+        this.cursor >= (this.currentPage?.value.length ?? 0)
+      ) {
+        this.complete = true;
+        this.pagingState = "Complete";
+        return;
+      }
+
+      const nextPage = await this.next();
+      if (!nextPage) {
+        return;
+      }
+      if (!nextPage.odataNextLink && nextPage.odataDeltaLink) {
+        this.pagingState = "Delta";
```

Comment:
Won't the `Delta` state get overwritten on the next interation of the while loop when `enumerate` is called and the results of the page are processed? 

Maybe we can add an example test case below(the next page only contains results and a delta link(no next link))
https://github.com/microsoftgraph/msgraph-sdk-dotnet-core/blob/0b6e132bbf0400b26560c384f965e91be315f1bd/tests/Microsoft.Graph.DotnetCore.Core.Test/Tasks/PageIteratorTests.cs#L227